include/swift/RemoteInspection/TypeRefBuilder.h

```diff
@@ -567,62 +567,58 @@ class TypeRefBuilder {
     return nullptr;
   }
 
+  // Recursively constructs TypeRef for a bound generic, including
+  // the enclosing (parent) generic contexts
   const BoundGenericTypeRef *createBoundGenericTypeReconstructingParent(
       const NodePointer node, const TypeRefDecl &decl, size_t shapeIndex,
       const llvm::ArrayRef<const TypeRef *> &args, size_t argsIndex) {
     if (!node || !node->hasChildren())
       return nullptr;
-    
+    NodePointer parentNode = node->getFirstChild();
+
+    // If this node can't possibly be generic, skip it and
+    // return results for our parent instead
+    auto kind = node->getKind();
+    if (kind != Node::Kind::Class
+	&& kind != Node::Kind::Structure
+	&& kind != Node::Kind::Enum) {
+      return createBoundGenericTypeReconstructingParent(
+	parentNode, decl, --shapeIndex, args, argsIndex);
+    }
+
+    // How many generic args are at this level?
     auto maybeGenericParamsPerLevel = decl.genericParamsPerLevel;
     if (!maybeGenericParamsPerLevel)
       return nullptr;
-
     auto genericParamsPerLevel = *maybeGenericParamsPerLevel;
-
-    auto kind = node->getKind();
-    // Kinds who have a "BoundGeneric..." variant.
-    if (kind != Node::Kind::Class && kind != Node::Kind::Structure &&
-        kind != Node::Kind::Enum)
+    if (shapeIndex >= genericParamsPerLevel.size())
       return nullptr;
+    auto numGenericArgs = genericParamsPerLevel[shapeIndex];
+
+    // Nodes with no generic args can be replaced with their parent
+    if (numGenericArgs == 0) {
+      return createBoundGenericTypeReconstructingParent(
```

Comment:
I don't understand the logic here, why can we ignore the current node? From the lldb test failure, it looks like we're dropping the innermost non-generic type, maybe because of this?

Given:
```
class F {
  class G<T> {
    class H {
      ...
    }
  }
}
```

The typeref of `F.G<Bool>.H` is being reconstructed as F.G<Bool>

Reply:
I think it's because a non-generic nested inside a generic one is actually generic, but not the other way around (which was my original mistake)?

So:

```
class F { // F is not generic.
  class G<T> { // G is generic.
    class H { // H is generic.
      ...
    }
  }
}
```

Reply:
Oh, good catch.  Yeah, I mixed something up here.  There's a basic difference between the original type we're constructing a reference for (which we have to reconstruct even if it's not a bound generic) and the parent types (where we only care about the bound generics).  That difference isn't being reflected in the code here.

The tricky part:  I don't think "bound generic" can mean "relies on generic information."  It has to mean "has generic type parameters."   So in your code above, we should not treat `H` as a "bound generic."  (Otherwise, `H` is ambiguous; we'd have to treat it as a bound generic when evaluated directly, but not when we evaluate another generic type contained inside it.)  But you're right that we have to construct a TypeRef for the type we're currently examining regardless of whether it's a bound generic or not.

Reply:
I think I've now correctly reworked things to handle this case correctly as well.  Basically, we now include the innermost typeref regardless (since that's the type we're actually trying to build a typeref for), but only include parent nodes that have non-zero numbers of generic type parameters.  I also added a test to check the case where the innermost type has no generic parameters.